staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark_test.go

```diff
@@ -94,28 +97,26 @@ func TestSampler(t *testing.T) {
 	}
 }
 
-/* getHistogramCount returns the count of the named histogram */
-func getHistogramCount(regs Registerables, metricName string) (int64, error) {
-	considered := []string{}
-	mfs, err := legacyregistry.DefaultGatherer.Gather()
+/* getHistogramCount returns the count of the named histogram or an error (if any) */
+func getHistogramCount(registry compbasemetrics.KubeRegistry, metricName string) (int64, error) {
+	mfs, err := registry.Gather()
 	if err != nil {
-		return 0, fmt.Errorf("failed to gather metrics: %s", err)
+		return 0, fmt.Errorf("failed to gather metrics: %w", err)
 	}
 	for _, mf := range mfs {
 		thisName := mf.GetName()
 		if thisName != metricName {
-			considered = append(considered, thisName)
 			continue
 		}
 		metric := mf.GetMetric()[0]
 		hist := metric.GetHistogram()
 		if hist == nil {
-			return 0, fmt.Errorf("dto.Metric has nil Histogram")
+			return 0, errors.New("dto.Metric has nil Histogram")
 		}
 		if hist.SampleCount == nil {
-			return 0, fmt.Errorf("dto.Histogram has nil SampleCount")
+			return 0, errors.New("dto.Histogram has nil SampleCount")
 		}
 		return int64(*hist.SampleCount), nil
 	}
-	return 0, fmt.Errorf("not found, considered=%#+v", considered)
+	return 0, errMetricNotFound
```

Comment:
why return an error, if it's not found, then 0 is the correct count.

Reply:
Because there are two ways to get zero: metric not found, or metric found and contains zero.  No need to lose that distinction.

Reply:
I actually don't think the metric exists until it gets written to...

Reply:
Yes, that is part of what is expected.
My point here is let's not needlessly discard a bit of information about _why_ the get method returned zero, since the point of a test is to not assume that everything goes as expected.

Reply:
If a metric can't exist until it's written to, then the error condition is actually not here, it exists in like 119. If `int64(*hist.SampleCount)` is equal to zero, this is a condition we do not expect and that should be an error.

Reply:
I am having trouble parsing "... is actually not here, it exists ...".

This is a behavioral unit test of the sample-and-watermark histograms including their underlying machinery.  While us developers _expect_ that the HistogramVec has no metrics before it is written, the point of a behavioral unit test is to not assume more than is necessary.  The current revision of this PR can distinguish between different pathologies that lead to an unexpected zero.  That seems better to me than not helping to identify what went wrong, in the case of an unexpected zero.

Reply:
> While us developers expect that the HistogramVec has no metrics before it is written...

This is a reasonable expectation given that this is how underlying Prometheus implementation actually works.

I am saying this is how it should look:

```
	for _, mf := range mfs {
		thisName := mf.GetName()
		if thisName != metricName {
			continue
		}
		metric := mf.GetMetric()[0]
		hist := metric.GetHistogram()
		if hist == nil {
			return 0, errors.New("dto.Metric has nil Histogram")
		}
		if hist.SampleCount == nil {
			return 0, errors.New("dto.Histogram has nil SampleCount")
		}
        count := int64(*hist.SampleCount)
        if count == 0 {
            return 0, errors.New("we should never have a 0 samplecount here")
        }
		return count, nil
	}
	return 0, nil
```

Reply:
I think it is unnecessarily specific for this client of the Prometheus go library to insist that a HistogramVec whose label slice is empty start out in a state where the suggested code in the previous comment executes the `return 0, nil` statement.  Remember that calling NewHistogram produces a Histogram with a sample count of zero.  So such a thing is perfectly fine, semantically.  A HistogramVec whose label slice is empty can only ever have one Histogram in it.  If the HistogramVec implementation were to choose to create the only possible Histogram in this case eagerly, who cares?  Maybe somebody with broader Prometheus concerns in mind, but I do not think that clients of sample-and-watermark histograms would care.